internal/cache/connector/redis/connector.go

```diff
@@ -0,0 +1,154 @@
+	DialTimeout time.Duration
+	// Timeout for socket reads. If reached, commands will fail
+	// with a timeout instead of blocking. Supported values:
+	//   - `0` - default timeout (3 seconds).
+	//   - `-1` - no timeout (block indefinitely).
+	//   - `-2` - disables SetReadDeadline calls completely.
+	ReadTimeout time.Duration
+	// Timeout for socket writes. If reached, commands will fail
+	// with a timeout instead of blocking.  Supported values:
+	//   - `0` - default timeout (3 seconds).
+	//   - `-1` - no timeout (block indefinitely).
+	//   - `-2` - disables SetWriteDeadline calls completely.
+	WriteTimeout time.Duration
+
+	// Type of connection pool.
+	// true for FIFO pool, false for LIFO pool.
+	// Note that FIFO has slightly higher overhead compared to LIFO,
+	// but it helps closing idle connections faster reducing the pool size.
+	PoolFIFO bool
+	// Base number of socket connections.
+	// Default is 10 connections per every available CPU as reported by runtime.GOMAXPROCS.
+	// If there is not enough connections in the pool, new connections will be allocated in excess of PoolSize,
+	// you can limit it through MaxActiveConns
+	PoolSize int
+	// Amount of time client waits for connection if all connections
+	// are busy before returning an error.
+	// Default is ReadTimeout + 1 second.
+	PoolTimeout time.Duration
+	// Minimum number of idle connections which is useful when establishing
+	// new connection is slow.
+	// Default is 0. the idle connections are not closed by default.
+	MinIdleConns int
+	// Maximum number of idle connections.
+	// Default is 0. the idle connections are not closed by default.
+	MaxIdleConns int
+	// Maximum number of connections allocated by the pool at a given time.
+	// When zero, there is no limit on the number of connections in the pool.
+	MaxActiveConns int
+	// ConnMaxIdleTime is the maximum amount of time a connection may be idle.
+	// Should be less than server's timeout.
+	//
+	// Expired connections may be closed lazily before reuse.
+	// If d <= 0, connections are not closed due to a connection's idle time.
+	//
+	// Default is 30 minutes. -1 disables idle timeout check.
+	ConnMaxIdleTime time.Duration
+	// ConnMaxLifetime is the maximum amount of time a connection may be reused.
+	//
+	// Expired connections may be closed lazily before reuse.
+	// If <= 0, connections are not closed due to a connection's age.
+	//
+	// Default is to not close idle connections.
+	ConnMaxLifetime time.Duration
+
+	EnableTLS bool
+
+	// Disable set-lib on connect. Default is false.
+	DisableIndentity bool
+
+	// Add suffix to client name. Default is empty.
+	IdentitySuffix string
+}
+
+type Connector struct {
+	*redis.Client
+	Config Config
+}
+
+func NewConnector(config Config) *Connector {
+	if !config.Enabled {
+		return nil
+	}
+	return &Connector{
+		Client: redis.NewClient(optionsFromConfig(config)),
+		Config: config,
+	}
+}
+
+func optionsFromConfig(c Config) *redis.Options {
+	opts := &redis.Options{
```

Comment:
The `DB` field is onw of the reasons why I do not use the `redis.Options`. Because a redis DB has a single keyspace, it is impossible to have multiple object caches on the same DB. If we would use any single DB, we would need to prefix each key with its "purpose". However, when doing the `Truncate` for a particular cache we would need to scan using the [`KEYS` command](https://redis.io/docs/latest/commands/keys/) with a regex. This is not recommended in production.

Instead, I provided the `DBOffset` option:

https://github.com/zitadel/zitadel/blob/66a61d07133017177b30c2983576856e93ed2e4d/internal/cache/connector/redis/connector.go#L29-L33

And we increment the used DB by 1, based on the Purpose enum:

https://github.com/zitadel/zitadel/blob/66a61d07133017177b30c2983576856e93ed2e4d/internal/cache/cache.go#L11-L19

Each script uses the [ `SELECT` command](https://redis.io/docs/latest/commands/select/) so that it manages the correct key space for it's object cache.

Reply:
Other reasons why we can't use `redis.Options` directly is:

- `tls.Config` is a complex type with function fields. We need to instantiate that ourselves and not let it be parsed from any YAML
-  The `Limiter` field is an interface used for rate limiting or Circuit breaking. (I'm still working on an implementation of the latter and will be a separate PR)